pkg/authorization/admission/restrictusers/restrictusers.go

```diff
@@ -0,0 +1,158 @@
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	oclient "github.com/openshift/origin/pkg/client"
+	oadmission "github.com/openshift/origin/pkg/cmd/server/admission"
+)
+
+func init() {
+	admission.RegisterPlugin("RestrictSubjectBindings",
+		func(kclient kclientset.Interface, config io.Reader) (admission.Interface,
+			error) {
+			return NewRestrictUsersAdmission(kclient)
+		})
+}
+
+// restrictUsersAdmission implements admission.Interface and enforces
+// restrictions on adding role bindings in a project to permit only designated
+// collaborators to add role bindings to a project.
+type restrictUsersAdmission struct {
+	*admission.Handler
+
+	oclient oclient.Interface
+	kclient kclientset.Interface
+}
+
+var _ = oadmission.WantsOpenshiftClient(&restrictUsersAdmission{})
+var _ = oadmission.Validator(&restrictUsersAdmission{})
+
+// NewRestrictUsersAdmission configures an admission plugin that enforces
+// restrictions on adding role bindings in a project based on predefined
+// collaborators.
+func NewRestrictUsersAdmission(kclient kclientset.Interface) (admission.Interface, error) {
+	return &restrictUsersAdmission{
+		Handler: admission.NewHandler(admission.Create),
+		kclient: kclient,
+	}, nil
+}
+
+func (q *restrictUsersAdmission) SetOpenshiftClient(c oclient.Interface) {
+	q.oclient = c
+}
+
+// Admit makes admission decisions that enforce restrictions on adding
+// project-scoped role-bindings.  In order for a role binding to be permitted,
+// each subject in the binding must be matched by some collaborator matcher in
+// the namespace.
+func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
+	if a.GetResource().GroupResource() != authorizationapi.Resource("rolebindings") {
+		return nil
+	}
+
+	// Ignore all operations that correspond to subresource actions.
+	if len(a.GetSubresource()) != 0 {
+		return nil
+	}
+
+	ns := a.GetNamespace()
+
+	// Ignore cluster-level resources.
+	if len(ns) == 0 {
+		return nil
+	}
+
+	binding, ok := a.GetObject().(*authorizationapi.RoleBinding)
+	if !ok {
+		return admission.NewForbidden(a,
+			fmt.Errorf("wrong object type for new rolebinding: %t", a.GetObject()))
+	}
+
+	oldbinding, ok := a.GetOldObject().(*authorizationapi.RoleBinding)
+	if oldbinding != nil && !ok {
+		return admission.NewForbidden(a,
+			fmt.Errorf("wrong object type for old rolebinding: %t", a.GetOldObject()))
+	}
+
+	glog.V(4).Infof("Handling rolebinding %s/%s", binding.Namespace, binding.Name)
+
+	// Only bother checking deltas.
+	newSubjects := []kapi.ObjectReference{}
+	if oldbinding == nil {
+		newSubjects = binding.Subjects
+	} else {
+		for _, newSubject := range binding.Subjects {
+			foundExisting := false
+			for _, existingSubject := range oldbinding.Subjects {
+				if existingSubject.Kind == newSubject.Kind &&
```

Comment:
ought to be able to just check straight `==` right?  We'll never false negative and false positive is ok.